sssd_test_framework/utils/virtual_smartcard.py

```diff
@@ -0,0 +1,134 @@
+from pytest_mh.cli import CLIBuilder
+
+__all__ = [
+    "SmartCardUtils",
+]
+
+
+class SmartCardUtils(MultihostUtility[MultihostHost]):
+    """
+    Utility class for managing smart card operations using SoftHSM and PKCS#11.
+    """
+
+    SOFTHSM2_CONF_PATH = "/opt/test_ca/softhsm2.conf"
+    TOKEN_STORAGE_PATH = "/opt/test_ca/tokens"
+    OPENSC_CACHE_PATHS = [
+        "$HOME/.cache/opensc/",
+        "/run/sssd/.cache/opensc/",
+    ]
+
+    def __init__(self, host: MultihostHost) -> None:
+        super().__init__(host)
+
+    def init(self, label: str, so_pin: str, user_pin: str) -> None:
+        """
+        Initializes a SoftHSM token with the given label and PINs.
+
+        Cleans cache directories and prepares the token directory.
+
+        :param label: Token label.
+        :param so_pin: Security Officer PIN.
+        :param user_pin: User PIN.
+        """
+        for path in self.OPENSC_CACHE_PATHS:
+            self.host.conn.run(f"rm -rf {path}")
+
+        self.host.conn.run(f"rm -rf {self.TOKEN_STORAGE_PATH}")
+        self.host.conn.run(f"mkdir -p {self.TOKEN_STORAGE_PATH}")
+
+        self.host.conn.exec(
+            [
+                "softhsm2-util", "--init-token",
+                "--label", label,
+                "--free",
+                "--so-pin", so_pin,
+                "--pin", user_pin
+            ],
+            env={"SOFTHSM2_CONF": self.SOFTHSM2_CONF_PATH}
+        )
+
+    def add_key(self, key_path: str, key_id: str = "01", pin: str = "123456") -> None:
+        """
+        Adds a private key to the smart card.
+
+        :param key_path: Path to the private key.
+        :param key_id: Key ID (default '01').
+        :param pin: User PIN (default '123456').
+        """
+        self.host.conn.exec(
+            [
+                "pkcs11-tool", "--module", "/usr/lib64/pkcs11/libsofthsm2.so",
+                "-l", "--pin", pin,
+                "--write-object", key_path,
+                "--type", "privkey",
+                "--id", key_id
+            ],
+            env={"SOFTHSM2_CONF": self.SOFTHSM2_CONF_PATH}
+        )
+
+    def add_cert(self, cert_path: str, cert_id: str = "01", pin: str = "123456") -> None:
+        """
+        Adds a certificate to the smart card.
+
+        :param cert_path: Path to the certificate.
+        :param cert_id: Certificate ID (default '01').
+        :param pin: User PIN (default '123456').
+        """
+        self.host.conn.exec(
+            [
+                "pkcs11-tool", "--module", "/usr/lib64/pkcs11/libsofthsm2.so",
+                "-l", "--pin", pin,
+                "--write-object", cert_path,
+                "--type", "cert",
+                "--id", cert_id
+            ],
+            env={"SOFTHSM2_CONF": self.SOFTHSM2_CONF_PATH}
+        )
+
+    def reset_service(self) -> None:
+        """
+        Restarts the virtual smart card service.
+        """
+        self.host.svc.restart("virt_cacard.service")
```

Comment:
I get a  mypy warning here:
```
sssd_test_framework/utils/virtual_smartcard.py:104: error: "MultihostHost[Any]" has no attribute "svc"  [attr-defined]
```
I think we need to add svc to the class so we can use it properly.   It can be setup like it is in the utils/sssd.py util class.   It also needs to be added to client to pass self.svc along with self.fs.